### DISCUSSION TITLE
Error: Please set rune path for node index 1 in RTL-Config.json!

### DISCUSSION THREAD

**Discussion by bitecorner** [Ride-The-Lightning/c-lightning-REST#213] [Q&A]
Hello,

When I try to run 
node rtl under user rtl:rtl, no root access
I placed the macaroon file in home rtl and write the path in the config file as follow
my lightning node running under lightning:lightning
I am facing this issue about the rune path ???
Any idea what i did wrong ? and how i could fix this ?
Bets regards
Jorge.

Here is my config file RTL-Config.json as follow :
{
"MultiPass": "XXXXXXXXX",
  "port": "3000",
  "defaultNodeIndex": 1,
  "SSO": {
    "rtlSSO": 0,
    "rtlCookiePath": "",
    "logoutRedirectLink": ""
  },
  "nodes": [
    {
      "index": 1,
      "lnNode": "Core_Lightning",
      "lnImplementation": "CLN",
      "Authentication": {
        "macaroonPath": "/home/rtl",
        "configPath": "/home/lightning/.lightning/config"
      },
      "Settings": {
        "userPersona": "OPERATOR",
        "themeMode": "NIGHT",
        "themeColor": "PURPLE",
        "channelBackupPath": "/home/rtl/backup-rtl",
        "logLevel": "ERROR",
        "lnServerUrl": "https://localhost:3001",
        "fiatConversion": false
      }
    }
  ],
  "multiPassHashed": "4634dad7bee28426e7612593f4d5bc64bf5171eb66ebd670c61fb5f0ad85e7fa"
}

**Comment by bitecorner** [comment]
tail -n 20 syslog
Feb  9 09:02:59 Nostrserver systemd[1]: Stopped Ride the Lightning.
Feb  9 09:02:59 Nostrserver systemd[1]: rtl.service: Consumed 1.386s CPU time.
Feb  9 09:02:59 Nostrserver systemd[1]: Started Ride the Lightning.
Feb  9 09:03:00 Nostrserver node[27803]: [2/9/2024, 9:03:00 AM] ERROR: Config => Something went wrong while configuring the node server:
Feb  9 09:03:00 Nostrserver node[27803]: Error: Please set rune path for node index 1 in RTL-Config.json!.
Feb  9 09:03:00 Nostrserver node[27803]: file:///home/rtl/RTL/backend/utils/config.js:369
Feb  9 09:03:00 Nostrserver node[27803]:                 throw new Error(err);
Feb  9 09:03:00 Nostrserver node[27803]:                       ^
Feb  9 09:03:00 Nostrserver node[27803]: Error: Error: Please set rune path for node index 1 in RTL-Config.json!
Feb  9 09:03:00 Nostrserver node[27803]:     at ConfigService.setServerConfiguration (file:///home/rtl/RTL/backend/utils/config.js:369:23)
Feb  9 09:03:00 Nostrserver node[27803]:     at new ConfigService (file:///home/rtl/RTL/backend/utils/config.js:372:14)
Feb  9 09:03:00 Nostrserver node[27803]:     at file:///home/rtl/RTL/backend/utils/config.js:375:23
Feb  9 09:03:00 Nostrserver node[27803]:     at ModuleJob.run (node:internal/modules/esm/module_job:218:25)
Feb  9 09:03:00 Nostrserver node[27803]:     at async ModuleLoader.import (node:internal/modules/esm/loader:323:24)
Feb  9 09:03:00 Nostrserver node[27803]:     at async loadESM (node:internal/process/esm_loader:28:7)
Feb  9 09:03:00 Nostrserver node[27803]:     at async handleMainPromise (node:internal/modules/run_main:120:12)
Feb  9 09:03:00 Nostrserver node[27803]: Node.js v21.6.1
Feb  9 09:03:00 Nostrserver systemd[1]: rtl.service: Main process exited, code=exited, status=1/FAILURE
Feb  9 09:03:00 Nostrserver systemd[1]: rtl.service: Failed with result 'exit-code'.
Feb  9 09:03:00 Nostrserver systemd[1]: rtl.service: Consumed 1.372s CPU time.

**Comment by ShahanaFarooqui** [comment]
@bitecorner Please note that RTL 0.15.0 is a major update which migrated RTL from c-lightning-REST to clnrest plugin. It will require working clnrest plugin on core lightning, a valid rune saved in the file and RTL-Config changes with rune details. 

https://github.com/Ride-The-Lightning/RTL/releases/tag/v0.15.0

**Comment by n-e-t-d-i-v-e-r** [comment]
hi there
same issue here:

`{
  "port": "9090",
  "SSO": {
    "rtlSSO": 0,
    "rtlCookiePath": "",
    "logoutRedirectLink": ""
  },
  "nodes": [
    {
      "index": 1,
      "lnNode": "CLN Node",
      "lnImplementation": "CLN",
      "Authentication": {
        "macaroonPath": "/home/rtl",
        "configPath": "/data/lightningd/config"
      },
      "Settings": {
        "userPersona": "OPERATOR",
        "themeMode": "DAY",
        "themeColor": "INDIGO",
        "fiatConversion": true,
        "currencyUnit": "EUR",
        "logLevel": "ERROR",
        "lnServerUrl": "http://127.0.0.1:3092",
        "enableOffers": true
      }
    }
  ],
  "defaultNodeIndex": 1,
  "multiPassHashed": ""
}`


[31/05/2024, 19:44:18] ERROR: Config => Something went wrong while configuring the node server: 
Error: Please set rune path for node index 1 in RTL-Config.json!.

file:///home/rtl/RTL/backend/utils/config.js:369
                throw new Error(err);
                      ^

Error: Error: Please set rune path for node index 1 in RTL-Config.json!
    at ConfigService.setServerConfiguration (file:///home/rtl/RTL/backend/utils/config.js:369:23)
    at new ConfigService (file:///home/rtl/RTL/backend/utils/config.js:372:14)
    at file:///home/rtl/RTL/backend/utils/config.js:375:23
    at ModuleJob.run (node:internal/modules/esm/module_job:222:25)
    at async ModuleLoader.import (node:internal/modules/esm/loader:316:24)
    at async asyncRunEntryPointWithESMLoader (node:internal/modules/run_main:123:5)

Node.js v20.13.1

### Should I fix a path in this json? If yes, where and how?


`export class ConfigService {
    constructor() {
        this.platform = os.platform();
        this.hash = crypto.createHash('sha256');
        this.errMsg = '';
        this.directoryName = dirname(fileURLToPath(import.meta.url));
        this.common = Common;
        this.logger = Logger;
        this.setDefaultConfig = () => {
            const homeDir = os.userInfo().homedir;
            let macaroonPath = '';
            let configPath = '';
            let channelBackupPath = '';
            let dbPath = '';
            switch (this.platform) {
                case 'win32':
                    macaroonPath = homeDir + '\\AppData\\Local\\Lnd\\data\\chain\\bitcoin\\mainnet';
                    configPath = homeDir + '\\AppData\\Local\\Lnd\\lnd.conf';
                    channelBackupPath = homeDir + '\\backup\\node-1';
                    dbPath = homeDir + '\\database\\node-1';
                    break;
                case 'darwin':
                    macaroonPath = homeDir + '/Library/Application Support/Lnd/data/chain/bitcoin/mainnet';
                    configPath = homeDir + '/Library/Application Support/Lnd/lnd.conf';
                    channelBackupPath = homeDir + '/backup/node-1';
                    dbPath = homeDir + '/database/node-1';
                    break;
                case 'linux':
                    macaroonPath = homeDir + '/.lnd/data/chain/bitcoin/mainnet';
                    configPath = homeDir + '/.lnd/lnd.conf';
                    channelBackupPath = homeDir + '/backup/node-1';
                    dbPath = homeDir + '/database/node-1';
                    break;
                default:
                    macaroonPath = '';
                    configPath = '';
                    channelBackupPath = '';
                    dbPath = '';
                    break;
            }
            const configData = {
                port: '3000',
                defaultNodeIndex: 1,
                dbDirectoryPath: dbPath,
                SSO: {
                    rtlSSO: 0,
                    rtlCookiePath: '',
                    logoutRedirectLink: ''
                },
                nodes: [
                    {
                        index: 1,
                        lnNode: 'Node 1',
                        lnImplementation: 'LND',
                        Authentication: {
                            macaroonPath: macaroonPath,
                            configPath: configPath
                        },
                        Settings: {
                            userPersona: 'MERCHANT',
                            themeMode: 'DAY',
                            themeColor: 'PURPLE',
                            channelBackupPath: channelBackupPath,
                            logLevel: 'ERROR',
                            lnServerUrl: 'https://127.0.0.1:8080',
                            fiatConversion: false,
                            unannouncedChannels: false
                        }
                    }
                ]
            };
            if ((process?.env?.RTL_SSO && +process?.env?.RTL_SSO === 0) || configData.SSO.rtlSSO === 0) {
                configData['multiPass'] = 'password';
            }
            return configData;
        };`

**Comment by ShahanaFarooqui** [comment]
@n-e-t-d-i-v-e-r Hey, the solution is also same :). 

RTL's v0.15.0 release is a major integration upgrade and a breaking change for its integration with Core Lightning.
With this release c-lightning-rest integration has been removed and RTL now uses core lightning’s native CLNRest plugin to communicate with the core lightning backend.

You can check more details in release notes at [0.15.0](https://github.com/Ride-The-Lightning/RTL/releases/tag/v0.15.0).

To know more about clnrest and how to enable it on CLN, you can check [clnrest documentation](https://docs.corelightning.org/docs/rest).

Finally once the clnrest is working on your core lightning node, you need to:

1: Create a new [rune](https://docs.corelightning.org/reference/lightning-createrune) or use an already existing one.

2: Create a new file where the content of the file must be LIGHTNING_RUNE="<your-rune-from-step-1>".

3: Update your lnServerUrl with the clnrest url.

4: Remove `macaroonPath` from RTL-Config.json for your CLN node.

5: Add `runePath` with path of the folder including filename which contains the rune for the node (from step 2).

**Comment by n-e-t-d-i-v-e-r** [comment]
@ShahanaFarooqui thank you for your quick response. as if I’m not a coder, I hope to get it right implemented into my raspibolt. usually I used this well documented manual: https://raspibolt.org

maybe I’ll do a pull-request there, in case my implementation along your description will be successful. :0)

but at least I wonder about this «rune»-label. is this the same technology that is used to write these stupid “NFTs” into the timechain, except that this is a useful use case? i publish the EINUNDZWANZIG-magazine, so i would be interested in your answer. ;0)

**Comment by n-e-t-d-i-v-e-r** [comment]
and btw, I guess you mean this rune-variant, right?


<img width="782" alt="Screenshot 2024-06-01 at 17 32 45" src="https://github.com/Ride-The-Lightning/c-lightning-REST/assets/87816768/3dba901a-20c2-4fe2-be16-325939ade0d7">

2: Create a new file where the content of the file must be LIGHTNING_RUNE="".
3: Update your lnServerUrl with the clnrest url.

I’m not sure with step 2 and 3. in which directory should I create that new file and with which naming of the file? the content of that file is without "unique_id"?

I guess in the RTL-directory, right?

<img width="507" alt="Screenshot 2024-06-01 at 17 53 18" src="https://github.com/Ride-The-Lightning/c-lightning-REST/assets/87816768/28425b14-83fc-4ff6-aa4b-3783dcbc51a1">

and I should change the «lnServerUrl» in the rtl-config.json, right? but where will I get this «clnrest url»?

**Reply by ShahanaFarooqui** [reply]
> and btw, I guess you mean this rune-variant, right?
> 
> <img alt="Screenshot 2024-06-01 at 17 32 45" width="782" src="https://private-user-images.githubusercontent.com/87816768/335811347-3dba901a-20c2-4fe2-be16-325939ade0d7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTcyOTA4NzAsIm5iZiI6MTcxNzI5MDU3MCwicGF0aCI6Ii84NzgxNjc2OC8zMzU4MTEzNDctM2RiYTkwMWEtMjBjMi00ZmUyLWJlMTYtMzI1OTM5YWRlMGQ3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjAyVDAxMDkzMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg4ZTUzZGZmY2RkNWYxNDM5ZmY2ODBmMTcxYmQyM2FkZjJlNzJhODE1MWE1NzdjMTdlZTBiM2Q5MzMzNzljOWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.VDIdLD3VoGhpvXIlsynuQD8-2cGzfrPWaRsJJxmU4uM">

Yes, we need a full access rune for RTL currently.

> 2: Create a new file where the content of the file must be LIGHTNING_RUNE="". 3: Update your lnServerUrl with the clnrest url.
> 
> I’m not sure with step 2 and 3. in which directory should I create that new file and with which naming of the file? the content of that file is without "unique_id"?
> 

For Step 2: You can save the file anywhere on your system with any name including RTL directory. For example, you can save the file inside `.lightning` directory with the name `.COMMANDO`. And it's content should be (no unique id, only rune):

```
LIGHTNING_RUNE="KUhZzNlECC7pYsz3QVbF1TqjIUYi3oyESTI7n60hLMs9MA=="
```

> and I should change the «lnServerUrl» in the rtl-config.json, right? but where will I get this «clnrest url»?

For Step 3: Yes, lnServerUrl needs to be updated in RTL-Config.json. If you have successfully followed through [clnrest documentation](https://docs.corelightning.org/docs/rest) and clnrest is working on your core lightning node, then you can find the url in core lighting logs, like:

`plugin-clnrest.py: REST Server is starting at https://127.0.0.1:3001`

or you can create the url by merging `<clnrest-protocol>://<clnrest-host>:<clnrest-port>` config values together. For example if my clnrest configurations are:
```
clnrest-port=3001
clnrest-protocol=https
clnrest-host=127.0.0.1
```
then the lnServerUrl would be `https://127.0.0.1:3001`.

**Comment by n-e-t-d-i-v-e-r** [comment]
sry, it’s me again. but if I’ll start RTL with «node rtl», I’ll get this ERROR:

<img width="719" alt="Screenshot 2024-06-02 at 09 04 07" src="https://github.com/Ride-The-Lightning/c-lightning-REST/assets/87816768/2879ae2f-0679-4241-b998-2d98bc7dfd8d">

my config looks like this:

<img width="540" alt="Screenshot 2024-06-02 at 09 04 40" src="https://github.com/Ride-The-Lightning/c-lightning-REST/assets/87816768/4e435ad2-e49f-446f-a4e3-d63bb23ccd05">

<img width="483" alt="Screenshot 2024-06-02 at 09 05 06" src="https://github.com/Ride-The-Lightning/c-lightning-REST/assets/87816768/99f134ba-ec4b-427c-b39a-31562d8f5bea">

<img width="747" alt="Screenshot 2024-06-02 at 09 12 16" src="https://github.com/Ride-The-Lightning/c-lightning-REST/assets/87816768/df600978-5362-46cc-ab09-6b4c51ecf32a">

<img width="453" alt="Screenshot 2024-06-02 at 08 52 44" src="https://github.com/Ride-The-Lightning/c-lightning-REST/assets/87816768/f022c6fd-a0ef-4708-bc5b-6d824af5c1ac">

and thet lnServerURL is like:
https://127.0.0.1:3092

**Comment by ShahanaFarooqui** [comment]
1: your .rune file should not have { and }. 

2: json doesn’t allow comments, so your macaroonPath will make it invalid.

3: you are not running clnrest which is an inbuilt Core Lightning plugin. You are running c-lightning-REST plugin which is a different plugin and is incompatible with RTL >=0.15.0 version. Please follow clnrest documentation for more details.

**Reply by n-e-t-d-i-v-e-r** [reply]
thx for the hint. but the problem is, I didn’t have an idea how to install clnrest. 

is there anywhere a step by step manual, how to install clnrest, like that?
https://raspibolt.org/guide/bonus/lightning/cln.html#c-lightning-rest--rtl

**Reply by ShahanaFarooqui** [reply]
I thought you wanted to open a manual PR for raspibolt :). 

As discussed earlier, for all clnrest related information (including [installation](https://docs.corelightning.org/docs/rest#installation)) is available on the link shared above.
